Introduce workload prioritization for service levels

piodul opened the pull request:
This series introduces workload prioritization: an extension of the service levels feature which allows specifying "shares" per service level. The number of shares determines the priority of the user which has this service level attached (if multiple are attached then the one with the lowest shares wins).

Different service levels will be isolated in the following way:

- Each service level gets its own scheduling group with the number of shares (corresponding to the service level's number of shares), which controls the priority of the CPU and I/O used for user operations running on that service level.
- Each service level gets two reader concurrency semaphores, one for user reads and the other for read-before-write done for view updates.
- Each service level gets its own TCP connections for RPC to prevent priority inversion issues.

Because of the mandatory use of scheduling groups, which are a globally limited resource, the number of service levels is now limited to 7 user created service levels + 1 created by default that cannot be removed.

This feature has been previously only available in ScyllaDB Enterprise but has been made available for the source available ScyllaDB. The series was created by comparing the master branch with source-available-workbranch / enterprise branch and taking the workload prioritization related parts from the diff, then molding the resulting diff into a proper series. Some very minor changes were made such as fixing whitespace, removing unused or unnecessary code, adding some boilerplate (in api/) which was missing, but otherwise no major changes have been made.

No backport is required.

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/22031)

**Changed Files:**
- [alternator/compatibility](http://previews.docs.scylladb.com/scylladb/22031/alternator/compatibility)
- [features/index.html](http://previews.docs.scylladb.com/scylladb/22031/features/index.html)
- [features/workload-attributes](http://previews.docs.scylladb.com/scylladb/22031/features/workload-attributes)
- [features/workload-prioritization](http://previews.docs.scylladb.com/scylladb/22031/features/workload-prioritization)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

tchaikov commented:
> [Test Result](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/testReport/) (2 failures / +2)
> 
>     [FullDtest / full-split006 / update_cluster_layout_tests.TestUpdateClusterLayout.test_simple_removenode_4](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split006___test_simple_removenode_4/)
>     [FullDtest / full-split006 / update_cluster_layout_tests.TestUpdateClusterLayout.test_simple_removenode_4](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split006___test_simple_removenode_4_2/)

#20814

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14048/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14048/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_metrics
:small_blue_diamond: alternator/test_service_levels
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/reader_concurrency_semaphore_test
:small_blue_diamond: boost/service_level_controller_test
:small_blue_diamond: topology_experimental_raft/test_alternator
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5606/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2490/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/267/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/863/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14048/)
#### Failed Tests (5/65749):
- [scylla_gdb.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.1)
- [scylla_gdb.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.2)
- [scylla_gdb.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.3)
- [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
- [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
#### Build Details:
- Duration: 5 hr 32 min
- Builder: spider7.cloudius-systems.com

piodul commented:
> * [scylla_gdb.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.1)
> * [scylla_gdb.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.2)
> * [scylla_gdb.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.3)

Looks like I forgot to update scylla-gdb.py in the commit that introduces the reader concurrency semaphore group. I will update the commit.

> * [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
> * [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14048/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)

I believe those are https://github.com/scylladb/scylladb/issues/20814

piodul commented:
> > [Test Result](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/testReport/) (2 failures / +2)
> > ```
> > [FullDtest / full-split006 / update_cluster_layout_tests.TestUpdateClusterLayout.test_simple_removenode_4](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split006___test_simple_removenode_4/)
> > [FullDtest / full-split006 / update_cluster_layout_tests.TestUpdateClusterLayout.test_simple_removenode_4](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13619/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split006___test_simple_removenode_4_2/)
> > ```
> 
> #20814

Ah, thanks @tchaikov! I didn't notice your comment at first.

piodul commented:
v2: updated scylla-gdb.py in the commit that adds reader concurrency semaphore groups

fruch commented:
@piodul I've run manually the relevant tests in dtest for this feature, and 4 of them are failing, TestSLATimeouts and test_sla_configuration

one looks like it expect the service level to appear in some parts of the tracing, and doesn't find them.
the other I don't know exactly what going on

I would recommend looking into those before merging this PR


```diff
diff --git a/sla_per_user_test.py b/sla_per_user_test.py
index 6e784491..c4156724 100644
--- a/sla_per_user_test.py
+++ b/sla_per_user_test.py
@@ -48,7 +48,6 @@ class SLATester(Tester):


 @pytest.mark.dtest_full
-@pytest.mark.dtest_enterprise
 @pytest.mark.single_node
 @pytest.mark.next_gating
 class TestSLA(SLATester):
@@ -382,7 +381,6 @@ class TestSLA(SLATester):


 @pytest.mark.dtest_full
-@pytest.mark.dtest_enterprise
 @pytest.mark.next_gating
 class TestSLANegativeTests(SLATester):
     def test_update_not_existing_sla(self):
@@ -564,7 +562,6 @@ class TestSLTimeoutsNegative(SLATester):
         assert exc.match(f".*{expected_exception_msg}.*")


-@pytest.mark.dtest_enterprise
 @pytest.mark.dtest_full
 class TestSLAConfig(SLATester):
     @staticmethod
```

```bash
SCYLLA_UNIFIED_PACKAGE=https://downloads.scylladb.com/unstable/scylla/master/relocatable/2024-12-23T22:19:57Z/scylla-unified-6.3.0~dev-0.20241223.6b0a657623a6.x86_64.tar.gz ./scripts/run_test.sh --scylla-version=sla_pr  sla_per_user_test.py
...
=========================== short test summary info ============================
FAILED sla_per_user_test.py::TestSLAConfig::test_sla_configuration - Assertio...
FAILED sla_per_user_test.py::TestSLATimeouts::test_timeout_valid_values[5h]
FAILED sla_per_user_test.py::TestSLATimeouts::test_timeout_valid_values[1000000ms]
FAILED sla_per_user_test.py::TestSLATimeouts::test_timeout_valid_values[23h2m2s2ms]
============ 4 failed, 23 passed, 5 deselected in 467.98s (0:07:47) ============
```

nyh commented on docs/operating-scylla/security/rbac-usecase.rst:
```diff
@@ -22,7 +22,7 @@ In the same manner, should someone leave the organization, all you would have to
    
 To build an RBAC environment, you need to create the roles and their associated permissions and then assign or grant the roles to the individual users. Roles inherit the permissions of any other roles that they are granted. The hierarchy of roles can be either simple or extremely complex. This gives great flexibility to database administrators, where they can  create specific permission condition ...
-In addition to standard roles, `ScyllaDB Enterprise <https://enterprise.docs.scylladb.com/>`_ users can implement `Workload Prioritization <https://enterprise.docs.scylladb.com/stable/using-scylla/workload-prioritization.html>`_, which allows you to attach roles to Service Levels, thus granting resources to roles as the role demands.
+In addition to standard roles, ScyllaDB Enterprise users can implement :doc:`Workload Prioritization </features/workload-prioritization/>`, which allows you to attach roles to Service Levels, thus granting resources to roles as the role demands.
```
Again, no need for the phrasing "ScyllaDB Enterprise" any more.

nyh commented on alternator/server.cc:
```diff
@@ -456,9 +456,16 @@ future<executor::request_return_type> server::handle_api_request(std::unique_ptr
+                co_return co_await callback(_executor, client_state, trace_state,
+                    make_service_permit(std::move(units)), std::move(json_request), std::move(req));
+    };
+    co_return co_await _sl_controller.with_user_service_level(user, std::ref(f));
```
Wow, this std::ref thing looks very odd. Why don't you just stuff the entire lambda as a parameter to the function?
Or maybe you just copied the existing code?

By the way, I hope that this with_user_service_level works - and efficiently - in the common (99% of the users) case where this service level feature isn't used. Even more importantly - it must also work (and do nothing) when there is NO user at all - when authentication is disabled.

piodul replied:
> Or maybe you just copied the existing code?

Yes, this part is taken directly from the enterprise repo (like most of the code in this PR). I believe this particular piece of logic was introduced in https://github.com/scylladb/scylla-enterprise/pull/3218.

I don't see an explanation as to why it was done like this, but given that `with_user_service_level` performs a `std::move` on the reference anyway I think I can safely do the transformation that you proposed.

> By the way, I hope that this with_user_service_level works - and efficiently - in the common (99% of the users) case where this service level feature isn't used.

With service levels on raft this is just a lookup in a `std::map`, so this should be pretty efficient.

> Even more importantly - it must also work (and do nothing) when there is NO user at all - when authentication is disabled.

In that case the lookup is skipped altogether, see the `else` branch in the `with_user_service_level` function.

nyh commented on test/alternator/conftest.py:
```diff
@@ -129,15 +128,15 @@ def dynamodb(request):
             config=boto_config.merge(botocore.client.Config(retries={"max_attempts": 0}, read_timeout=300)))
 
-def new_dynamodb_session(request, dynamodb):
+def new_dynamodb_session(request, dynamodb, user='cassandra', password='secret_pass'):
```
This change is wrong... The "password" isn't used below (?) and the "user" should be called "role".

nyh commented on test/alternator/run:
```diff
@@ -115,6 +115,19 @@ run.wait_for_services(pid, [
+# Additional role and service level are created to test the feature properly (alternator doesn't have it's own API to set it up so we need to use CQL).
+cql.execute("CREATE ROLE alternator_custom_sl WITH LOGIN = true AND HASHED PASSWORD = 'secret_pass'")
+cql.execute("CREATE SERVICE LEVEL sl_alternator")
+cql.execute("ATTACH SERVICE LEVEL sl_alternator TO alternator_custom_sl")
```
This is so sad :-( We had such code in the past for the default role, and no longer needed it, and it's sad we needed it back. In fact, I think it's a mistake to do it in test/alternator/run, since it can actually be done *inside* the test.  See for example test_cql_rbac.py where the test itself creates the role it needs for the test.
I don't expect you to do this now, it's ok that you merge this code now because this is what we had in Enterprise - but I really want to fix it so I opened https://github.com/scylladb/scylladb/issues/22047 and plan to fix this myself after this implementation goes in.

nyh commented on test/alternator/test_service_levels.py:
```diff
@@ -0,0 +1,95 @@
+def cql_table_name(tab):
+    return maybe_quote('alternator_' + tab.name) + '.' + maybe_quote(tab.name)
+
+# This file is all about testing RBAC as configured via CQL, so we need to
```
All this code is duplicate stuff we have in test_cql_rbac.py, so I added to https://github.com/scylladb/scylladb/issues/22047 the task to clean this duplication - I'm fine that you add this duplicate code for now.

The duplication is doubly silly when the comment says that "this file is all about testing RBAC", which clearly was copied from a file that was really about testing RBAC ;-)

nyh commented on test/alternator/test_service_levels.py:
```diff
@@ -0,0 +1,95 @@
+    timeout = 30
+    while True:
+        try:
+            with temporary_grant(cql, 'MODIFY', cql_table_name(test_table), 'alternator_custom_sl'):
```
This test did not need to use temporary_grant at all, it's not about checking RBAC... The test should have just used a superuser role which has permissions to do anything and not mess with RBAC! Again I'll add this cleaup task to the issue.

nyh replied:
I added this observation to #22047 so you you don't need to fix it now.

nyh commented on tracing/tracing.hh:
```diff
@@ -175,6 +175,7 @@ struct event_record {
     std::string message;
     elapsed_clock::duration elapsed;
     i_tracing_backend_helper::wall_clock::time_point event_time_point;
+    sstring scheduling_group_name = current_scheduling_group().name();
```
The commit message says it saves the "service level" but I see here the scheduling group name. I think we should be more consistent in the commit message (these can be different names, right?)

piodul replied:
This phrasing was used in the original commit that I cherry picked. Should I still change it?

nyh replied:
I don't know. I'm not even sure my complaint is correct. Do whatever you think is better. If there's an error in the commit message and you can spend 3 minutes to fix it, I think it's worth it.

nyh reviewed (commented):
Looks mostly good. I left some mostly trivial questions and comments. 
In one place where I had major gripes about the testing approach, I suggest that we accept this from-Enterprise code mostly unchanged and fix it later - I opened https://github.com/scylladb/scylladb/issues/22047 because it's really important for me to do this right (the test/alternator/run script is supposed to be a thin and even unnecessary runner, it shouldn't be a replacement for fixtures).

nyh commented on CMakeLists.txt:
```diff
@@ -92,7 +92,7 @@ else()
     set(Seastar_EXCLUDE_TESTS_FROM_ALL ON CACHE BOOL "" FORCE)
     set(Seastar_IO_URING ON CACHE BOOL "" FORCE)
-    set(Seastar_SCHEDULING_GROUPS_COUNT 16 CACHE STRING "" FORCE)
+    set(Seastar_SCHEDULING_GROUPS_COUNT 19 CACHE STRING "" FORCE)
```
I saw later that you do have a regression test for this "7 service levels" thing, so good.
This series is so longer I forgot to go back and remove this comment :-)

piodul replied:
> If it is documented that we "support 7 user-managed service levels"

For the record, this is documented for enterprise: https://enterprise.docs.scylladb.com/stable/using-scylla/workload-prioritization.html#limits. That documentation page is migrated in the commit titled `docs: add documentation for workload prioritization`.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14059/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14059/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_metrics
:small_blue_diamond: alternator/test_service_levels
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/reader_concurrency_semaphore_test
:small_blue_diamond: boost/service_level_controller_test
:small_blue_diamond: topology_experimental_raft/test_alternator
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/876/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5623/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13636/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2507/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/280/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14059/)
#### Build Details:
- Duration: 10 hr
- Builder: i-0ee5b29d476f0275d (m5ad.8xlarge)

piodul commented on db/system_distributed_keyspace.cc:
```diff
@@ -152,8 +155,14 @@ static const sstring CDC_TIMESTAMPS_KEY = "timestamps";
+        auto builder = schema_builder(system_distributed_keyspace::NAME, system_distributed_keyspace::SERVICE_LEVELS, std::make_optional(id))
                 .with_column("service_level", utf8_type, column_kind::partition_key)
+                .with_column("shares", int32_type);
+        if (utils::get_local_injector().is_enabled("service_levels_v1_table_without_shares")) {
```
Yes, the purpose of this error injection is to simulate an upgrade from a version which didn't support workload prioritization. It was introduced in scylladb/scylla-enterprise#4634.

For the record, we do have a proper upgrade test in dtests which does not do any error injections (enterprise_upgrade_test.py::TestEnterpriseUpgradeFromOSS).

nyh replied:
So maybe we don't need this injection thing at all. These injections all over the code are really starting to become unsightly (and too frequently - hard to understand, and we'll really hate them if one day we want to refactor code containing injections). But I guess that if you're just copying code that already existed in Enterprise and had this injection, we don't have to rethink this old decision right now :-(

piodul replied:
I'd rather keep it for now because it does some checks that the dtest doesn't. I'm not a great fan of the error injection in this particular place either, I think it was mostly done for convenience as ./test.py doesn't support proper rolling upgrade tests (yet?). It would be great to unify those tests, make them do a "real" rolling upgrade and put them into one place, but that would be a larger effort than I'd like to do in this PR.

avikivity replied:
@nyh you are reviewing existing code that is being ported. Any changes should be applied after the merge, or we will diverge.

piodul commented on test/topology_experimental_raft/test_alternator.py:
```diff
@@ -140,7 +140,7 @@ async def get_cpu_metrics():
             metrics = await manager.metrics.query(ip)
             ms_streaming += metrics.get('scylla_scheduler_runtime_ms', {'group': 'streaming'})
-            ms_statement += metrics.get('scylla_scheduler_runtime_ms', {'group': 'statement'})
+            ms_statement += metrics.get('scylla_scheduler_runtime_ms', {'group': 'sl:default'})
```
The service level controller enforces a naming convention on the scheduling groups associated with service levels: `sl:<service level name>`. The "statement" scheduling group is reused as the scheduling group for the "default" service level, hence it is renamed to `sl:default`.

Moreover, the "default" service level is documented (https://enterprise.docs.scylladb.com/stable/using-scylla/workload-prioritization.html#using-workload-prioritization-with-your-application) as the service level used when no user-created service level applies (i.e. anonymous user or user without any SL assigned). Given how this is explained in the documentation, I think make sense for this scheduling group to now be named `sl:default`.

nyh replied:
I hope we don't have users, or a bazillion dtests, who rely on that old name "statement". I guess it won't be a problem because we've had this name "sl:default" in the Enterprise branch, and I don't remember anyone complaining.

piodul commented on main.cc:
```diff
@@ -1501,7 +1500,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
             // Delay listening messaging_service until gossip message handlers are registered
-            messaging.start(mscfg, scfg, creds, std::ref(feature_service), std::ref(gossip_address_map)).get();
+            messaging.start(std::ref(sl_controller), mscfg, scfg, creds, std::ref(feature_service), std::ref(gossip_address_map)).get();
```
> I know you're just copying code we already had in Enterprise, but this patch made me feel a bit uneasy. First, it seems you removed the "$user" "tenant", which I have vague memory we recently had problems with, and I hope it doesn't break with this patch.

It was the "$maintenance" tenant that we recently had problems with. As for the "$user" tenant, we do have code to handle it, see `messaging_service::scheduling_group_for_isolation_cookie`.

piodul replied:
> Second, before this series service levels in the master branch were some low-impact feature through which you can ask for different timeouts or something. After this patch, the service levels mean a very different thing - having multiple service levels now adds different scheduling groups, adds more connections, etc. I hope we will document the new implications of having service levels.

@tzach FYI (pinging just in case, but the cover letter should make it rather clear)

Just in case, I'll also leave a link (https://github.com/scylladb/scylla-enterprise/pull/4759#issuecomment-2405531598) to a summary of a discussion that we had some time ago with @avikivity; there was an idea to repurpose NULL shares to make service levels with NULL shares behave as in open source, but we decided against that (would be a breaking change for enterprise users, SLs are not considered popular enough in OSS to justify the work).

piodul commented on db/config.hh:
```diff
@@ -500,6 +500,8 @@ public:
     named_value<bool> enable_tablets;
     named_value<uint32_t> view_flow_control_delay_limit_in_ms;
 
+    bool create_service_levels_table = true; // Some upgrade tests don't want enterprise-only tables
```
I think you are right.

Actually, this flag is not used anywhere - but only because I forgot to migrate the only part of the code which set it to false (in schema_change_test.cc). The test passed despite that; I was wondering why (new columns were added to system_distributed_keyspace.service_levels table) but it looks like system distributed tables are not taken into account by the test since e26aa836a97b5921e3890d35fa9111fdc75d246f.

I'll drop the flag because it is not needed for anything anymore.

piodul commented on service/storage_service.cc:
```diff
@@ -2035,6 +2037,35 @@ future<> storage_service::join_topology(sharded<db::system_distributed_keyspace>
+        // the callback might be run immediately and it uses async methods, so the thread is needed
+        co_await seastar::async([&] {
+            _workload_prioritization_registration = _feature_service.workload_prioritization.when_enabled([&sys_dist_ks] () {
+                // since we are creating tables here and we wouldn't want to have a race condition
```
The comment was taken verbatim from the enterprise code, but I can edit it if it is confusing.

nyh replied:
I see. Then I guess it's not fair to complain to you :-) I just thought that "thundering herd" was the more appropriate term here.

piodul commented on service/qos/qos_common.hh:
```diff
@@ -90,9 +91,9 @@ using service_levels_info = std::map<sstring, service_level_options>;
 /// A logical argument error for a service_level statement operation.
 ///
-class service_level_argument_exception : public std::invalid_argument {
+class service_level_argument_exception : public exceptions::invalid_request_exception {
```
Yes, I agree that this isn't great, the `service_level_argument_exception` belongs to a different abstraction layer than `exceptions::invalid_request_exception`. I copied this part from enterprise, it was originally introduced in https://github.com/scylladb/scylla-enterprise/commit/c802ee5d7903f8d0e17a3653228c7df25ca778ef.

> You are assuming that this exception can only happen from within a CQL request, so it makes sense to directly throw a CQL-specific exception. But is this assumption accurate? Couldn't we (at least a theory) get exactly the same exception in an Alternator request? In a REST API request?

I believe that this is an accurate assumption - this exception can only be thrown during a request to attach or drop (!) a service level, and AFAIK service levels cannot be manipulated in this way via Alternator, can they?

Actually, now I can see that our existing service levels code directly throws `invalid_request_exception`s... These should be converted into specific exceptions, too.

Considering that, I'd rather leave this to a follow-up to clean all of this up in one swoop. The base change from `std::invalid_argument` to `exceptions::invalid_request_exception` is most likely needed to make boost tests pass (I can see that those match on the exact exception type), so I'd rather include it in this PR - although flawed, this is a bugfix after all.

nyh replied:
Ok. Let's leave it like this and if it bothers us later, we'll fix it then. I didn't realize that this exception is only for the service-level modification operation, which indeed is (at least today and in the forseeable future) CQL-only.

piodul commented on cql3/statements/sl_prop_defs.hh:
```diff
@@ -18,8 +18,13 @@ namespace statements {
+    static constexpr auto KW_SHARES = "shares";
+    static constexpr int SHARES_DEFAULT_VAL = 1000;
+    static constexpr int SHARES_MIN_VAL = 1;
+    static constexpr int SHARES_MAX_VAL = 1000;
```
Unfortunately, I don't know the answer to your question. Perhaps @avikivity will remember or @Jadw1 will know?

nyh replied:
I'm sorry for asking all these questions now - I guess you copied most if not all of that stuff from the original code in the Enterprise branch. Maybe it's too late to ask these questions now, and we should just commit this code as-is. I'm not against this at all... But it's also an opportunity to (briefly) rethink things and improve (small) things, if we want.

avikivity replied:
The questions are fine, but if the answers are "a change is indicated" then the change should be a follow-on.

piodul commented on test/cqlpy/test_service_levels.py:
```diff
@@ -96,3 +113,63 @@ def test_list_effective_service_level(scylla_only, cql):
+        with pytest.raises(InvalidRequest, match=f"Role {role} doesn't have assigned any service level"):
+            cql.execute(f"LIST EFFECTIVE SERVICE LEVEL OF {role}")
+
+# Scylla Enterprise limits the number of service levels to a small number (8 including 1 default service level).
```
I'll go over all occurrences of Enterprise in this PR and will remove them.

tchaikov commented on reader_concurrency_semaphore_group.cc:
```diff
@@ -0,0 +1,112 @@
+
+#include "reader_concurrency_semaphore_group.hh"
+
+// Calling adjust is serialized since 2 adjustments can't happen simultaneosly,
```
could you please take this opportunity to fix the typo?

avikivity replied:
No, please do not fix typos. It will cause me pain when I merge back into enterprise.

The goal of the migrations is to make master == enterprise, not to fix typos, bugs, misfeatures, or anything.

tchaikov replied:
created #22154 to address these misspellings.

mykaul commented:
@piodul  - please resolve conflicts.

piodul commented on test/cqlpy/test_describe.py:
```diff
@@ -1488,19 +1487,28 @@ def __exit__(self, exc_type, exc_val, exc_tb) -> None:
 
 class ServiceLevel:
-    def __init__(self, name: str, timeout: int|None = None, wl_type: str|None = None):
+    default_shares_value = 1000
```
I'm pretty sure this one needs to be 1000, please see the `get_create_stmt` method and see the comment about the `replace_default_shares` parameter.

We use this method in DESCRIBE tests: when describing a service level, Scylla will include the `WITH shares = ...` part unconditionally, even if you didn't provide it when creating the service level, and in that case the number of shares will be equal to 1000. Our service levels tests check this property.

nyh replied:
But this is exactly my point - if your test always sets the value and then checks the value - why does it need to be 1000? It can be set to 987 and then reading it will retrieve 987, not 1000. The only place you should read 1000 is when you didn't override the default.

piodul replied:
It's not that the test always sets the shares, it's Scylla who writes the default shares into its internal tables. For example, if the client sends:

```
CREATE SERVICE LEVEL x;
```

...the DESCRIBE will return this statement:

```
CREATE SERVICE LEVEL x WITH shares = 1000;
```

Hence the `replace_default_shares` parameter - with `False` it generates the first variant, with `True` - the latter.

piodul commented on service/qos/service_level_controller.hh:
```diff
@@ -99,6 +99,8 @@ using update_both_cache_levels = bool_class<class update_both_cache_levels_tag>;
  */
 class service_level_controller : public peering_sharded_service<service_level_controller>, public service::endpoint_lifecycle_subscriber {
 public:
+    static inline const int32_t default_shares = 1000;
```
I thought about it for a bit and I don't think those are the same. There is:

- The default number of shares for a newly created service level, if the user doesn't specify the shares (this constant here)
- The number of shares for the default service level (defined via the default_service_level_config parameter of the service level controller's constructor)

They have indeed the same value, but I don't think they are conceptually the same.

I think I'd rather go in the direction of introducing some kind of a const global "default default (sic) service level config" with shares set to 1000, perhaps maybe even getting of default_service_level_config (it's only really used with custom values in boost tests, not sure yet why), but I'd rather avoid doing that in this PR, maybe in a followup.

avikivity replied:
the goal of this pull request is to reduce the differences with enterprise as much as possible.

avikivity commented:
Why is this 26 patches? I don't mind too much, but it's wasted effort. I explained that I'm ok with a coarse split here since this is reviewed code.

I also asked that the cover letter be phrased in a neutral way. Yes, we're porting it from enterprise (and it can be mentioned), but the main effect is introducing a new feature to master, and the subject should reflect that.

piodul commented:
v3:

- Removed config::create_service_levels_table
- Inlined lambda in alternator/server.cc, server::handle_api_request
- Fixed typo in comment in reader_concurrency_semaphore_group.cc
- Added information about service levels in docs/dev/isolation.md
- Updated documentation and comments not to mention Enterprise, if not needed
- Reworked comment about the sleep in service/storage_service.cc
- Cleaned up unused parameters / wrong names in test/alternator/conftest.py
- Got rid of unused imports in test/cqlpy/test_describe.py
- Updated the title and the cover letter to put more emphasis that it's a new feature and less that it was ported from enterprise

nyh commented:
> Why is this 26 patches? I don't mind too much, but it's wasted effort. I explained that I'm ok with a coarse split here since this is reviewed code.

I think there was some misunderstanding here. I took this pull request as a nice opportunity to review code that I never reviewed (perhaps you reviewed, I don't know), and appreciated @piodul's effort to make it reviewable. I didn't expect to do major changes in this review since we know this is already working code, but I hoped that we can get some small corrections, as well as discover bigger problems that need to be discovered later (e.g., I opened https://github.com/scylladb/scylladb/issues/22047). Since that @avikivity thinks this code doesn't really need to be reviewed again. This probably should have been decided earlier, not now, and other people doing these ports should be made aware of this as well.

avikivity commented:
> v3:
> 
> * Removed config::create_service_levels_table
> * Inlined lambda in alternator/server.cc, server::handle_api_request

This is really unfortunate. The plan for converging was to port a feature to master, merge back into enterprise, and keep going until the diff is nulled. Now the master->enterprise merge will be very difficult since the two sides aren't identical.

> * Fixed typo in comment in reader_concurrency_semaphore_group.cc
> * Added information about service levels in docs/dev/isolation.md
> * Updated documentation and comments not to mention Enterprise, if not needed
> * Reworked comment about the sleep in service/storage_service.cc
> * Cleaned up unused parameters / wrong names in test/alternator/conftest.py
> * Got rid of unused imports in test/cqlpy/test_describe.py
> * Updated the title and the cover letter to put more emphasis that it's a new feature and less that it was ported from enterprise

nyh commented:
There is a conflict in test/topology_experimental_raft/test_alternator.py, I guess just because in master everything from topology_experimental_raft moved to topology_custom.

nyh commented:
> > v3:
> > 
> > * Removed config::create_service_levels_table
> > * Inlined lambda in alternator/server.cc, server::handle_api_request
> 
> This is really unfortunate. The plan for converging was to port a feature to master, merge back into enterprise, and keep going until the diff is nulled. Now the master->enterprise merge will be very difficult since the two sides aren't identical.

I guess I am to blame for suggesting these improvements. I didn't understand that you are planning master->enterprise merges, and I still don't understand why they are needed: We don't need the "enterprise" head any more because we won't cut new enterprise releases from that branch! Backports to existing enterprise branches are done separately, we don't need the "enterprise" head for that.

It seems that neither I nor @piodul knew that your plan was NOT to review these enterprise-to-scylla.git porting pull-requests, and just merge them unchanged if the CI passes.

avikivity commented:
> > > v3:
> > > 
> > > * Removed config::create_service_levels_table
> > > * Inlined lambda in alternator/server.cc, server::handle_api_request
> > 
> > 
> > This is really unfortunate. The plan for converging was to port a feature to master, merge back into enterprise, and keep going until the diff is nulled. Now the master->enterprise merge will be very difficult since the two sides aren't identical.
> 
> I guess I am to blame for suggesting these improvements. I didn't understand that you are planning master->enterprise merges, and I still don't understand why they are needed: We don't need the "enterprise" head any more because we won't cut new enterprise releases from that branch! Backports to existing enterprise branches are done separately, we don't need the "enterprise" head for that.


My algorithm for ascertaining that nothing was missed during the porting is to keep merging master to enterprise until the diff is null. If changes are made during this process the merge becomes more difficult.

> 
> It seems that neither I nor @piodul knew that your plan was NOT to review these enterprise-to-scylla.git porting pull-requests, and just merge them unchanged if the CI passes.

piodul commented:
> > v3:
> > 
> > * Removed config::create_service_levels_table
> > * Inlined lambda in alternator/server.cc, server::handle_api_request
> 
> This is really unfortunate. The plan for converging was to port a feature to master, merge back into enterprise, and keep going until the diff is nulled. Now the master->enterprise merge will be very difficult since the two sides aren't identical.

@avikivity I can still bring back the previous version of the PR which was much closer to the enterprise branch. As for @nyh's suggestions, perhaps I could extract them to a separate PR and we could merge them as a follow-up?

What kinds of deviations from the enterprise branch are acceptable in this PR, if any? I suppose that we could abstain from small refactors / code cleanup, but what about e.g. mentions of Enterprise in the documentation and the code?

avikivity commented:
> > > v3:
> > > 
> > > * Removed config::create_service_levels_table
> > > * Inlined lambda in alternator/server.cc, server::handle_api_request
> > 
> > 
> > This is really unfortunate. The plan for converging was to port a feature to master, merge back into enterprise, and keep going until the diff is nulled. Now the master->enterprise merge will be very difficult since the two sides aren't identical.
> 
> @avikivity I can still bring back the previous version of the PR which was much closer to the enterprise branch.

Please do.

> As for @nyh's suggestions, perhaps I could extract them to a separate PR and we could merge them as a follow-up?

Yes, but only after the next master->enterprise merge.

> 
> What kinds of deviations from the enterprise branch are acceptable in this PR, if any? I suppose that we could abstain from small refactors / code cleanup, but what about e.g. mentions of Enterprise in the documentation and the code?

If it's just the word "enterprise" then we'll clean it up later. Private data should be removed, but I don't think the code contains any.

nyh commented:
> > As for @nyh's suggestions, perhaps I could extract them to a separate PR and we could merge them as a follow-up?
> 
> Yes, but only after the next master->enterprise merge.

From past experience, this will never happen. Nobody will remember to go back and do some old review comments a long time after the code was merged.

I propose that all these "enterprise" backports you should just merge automatically when CI passes, reviews like what I did (and spent many hours on) on these enterprise backports have become pointless.
 
> > What kinds of deviations from the enterprise branch are acceptable in this PR, if any? I suppose that we could abstain from small refactors / code cleanup, but what about e.g. mentions of Enterprise in the documentation and the code?
> 
> If it's just the word "enterprise" then we'll clean it up later. Private data should be removed, but I don't think the code contains any.

Again, I believe nobody will ever remember to go back to clean this up. I'll be happy to be disproved.

avikivity commented:
> > > As for @nyh's suggestions, perhaps I could extract them to a separate PR and we could merge them as a follow-up?
> > 
> > 
> > Yes, but only after the next master->enterprise merge.
> 
> From past experience, this will never happen. Nobody will remember to go back and do some old review comments a long time after the code was merged.
> 
> I propose that all these "enterprise" backports you should just merge automatically when CI passes, reviews like what I did (and spent many hours on) on these enterprise backports have become pointless.
> 
> > > What kinds of deviations from the enterprise branch are acceptable in this PR, if any? I suppose that we could abstain from small refactors / code cleanup, but what about e.g. mentions of Enterprise in the documentation and the code?
> > 
> > 
> > If it's just the word "enterprise" then we'll clean it up later. Private data should be removed, but I don't think the code contains any.
> 
> Again, I believe nobody will ever remember to go back to clean this up. I'll be happy to be disproved.

File an issue and assign it to @piodul

nyh commented:
> File an issue and assign it to @piodul

When I had a big gripe about something in the code that I knew @piodul will not fix right away (and shouldn't, I'll do it more easily myself), I opened https://github.com/scylladb/scylladb/issues/22047 - and I intend to do it later. But I didn't (and won't) open issues on the dozen smaller (sometimes even trivial) comments I made. That's not how our review process works normally - we have been using (in the last few years) Github's review UI to do these reviews, so this is what I did here too. I wasn't told I wasn't supposed to review the code, so I did :-(

piodul commented:
@avikivity Would it be fine if I submit a minimal cleanup PR to enterprise right now, in parallel to this PR? There is some code in enterprise which is either dead or is just unnecessarily more complex version of what we had in open source; it feels wrong for me to go and add it back, and not all of them would naturally fit in the existing commits of this PR, I'd like to avoid porting it right away.

avikivity commented:
Yes

piodul commented:
Cleanup PR in enterprise: https://github.com/scylladb/scylla-enterprise/pull/5117

piodul commented:
v4:
- Reverted the branch to an older version
- Submitted https://github.com/scylladb/scylla-enterprise/pull/5117 with a bunch of cleanups for the enterprise repo
- Diffed this PR's branch and https://github.com/scylladb/scylla-enterprise/pull/5117's branch and updated this PR until I didn't see any more major differences in service levels implementation between those branches

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14149/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14149/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_metrics
:small_blue_diamond: alternator/test_service_levels
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/reader_concurrency_semaphore_test
:small_blue_diamond: boost/service_level_controller_test
:small_blue_diamond: topology_custom/test_alternator
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/936/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2560/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5673/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13686/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/340/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14149/)
#### Build Details:
- Duration: 9 hr 16 min
- Builder: spider8.cloudius-systems.com

piodul commented:
v4.1: rebased onto current master

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14169/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14169/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_metrics
:small_blue_diamond: alternator/test_service_levels
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/reader_concurrency_semaphore_test
:small_blue_diamond: boost/service_level_controller_test
:small_blue_diamond: topology_custom/test_alternator
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/956/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2583/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5696/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13709/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/360/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14169/)
#### Build Details:
- Duration: 9 hr 19 min
- Builder: spider8.cloudius-systems.com
